buildSrc/src/main/kotlin/org/gradle/ShadedJarCreator.kt

```diff
@@ -0,0 +1,279 @@
+package org.gradle
+
+import org.gradle.internal.exceptions.Contextual
+
+import org.objectweb.asm.ClassReader
+import org.objectweb.asm.ClassWriter
+import org.objectweb.asm.commons.ClassRemapper
+import org.objectweb.asm.commons.Remapper
+
+import java.io.BufferedInputStream
+import java.io.BufferedOutputStream
+import java.io.File
+import java.io.FileInputStream
+import java.io.FileOutputStream
+import java.io.PrintWriter
+import java.net.URI
+import java.nio.file.Files
+import java.nio.file.FileSystems
+import java.nio.file.FileVisitResult
+import java.nio.file.SimpleFileVisitor
+import java.nio.file.Path
+import java.nio.file.attribute.BasicFileAttributes
+import java.util.jar.JarFile
+import java.util.jar.JarOutputStream
+import java.util.zip.ZipEntry
+
+
+private
+val ignoredPackagePatterns = PackagePatterns(setOf("java"))
+
+
+open class ShadedJarCreator(
+    private val sourceJars: Iterable<File>,
+    private val jarFile: File,
+    private val analysisFile: File,
+    private val classesDir: File,
+    private val shadowPackage: String,
+    private val keepPackages: Set<String>,
+    private val unshadedPackages: Set<String>,
+    private val ignorePackages: Set<String>) {
+
+    fun createJar() {
+        val start = System.currentTimeMillis()
+        PrintWriter(analysisFile).use { writer ->
+            val classes = classGraph()
+            analyse(classes, writer)
+            writeJar(classes, classesDir, jarFile, writer)
+        }
+        val end = System.currentTimeMillis()
+        println("Analysis took " + (end - start) + "ms.")
+    }
+
+    private
+    fun classGraph() =
+        ClassGraph(
+            PackagePatterns(keepPackages),
+            PackagePatterns(unshadedPackages),
+            PackagePatterns(ignorePackages),
+            shadowPackage)
+
+    private
+    fun analyse(classes: ClassGraph, writer: PrintWriter) =
+        sourceJars.forEach {
+            val jarUri = URI.create("jar:" + it.toPath().toUri())
+            FileSystems.newFileSystem(jarUri, emptyMap<String, Any>()).use { jarFileSystem ->
+                jarFileSystem.rootDirectories.forEach {
+                    visitClassDirectory(it, classes, ignoredPackagePatterns, writer)
+                }
+            }
+        }
+
+    private
+    fun visitClassDirectory(dir: Path, classes: ClassGraph, ignored: PackagePatterns, writer: PrintWriter) {
+        Files.walkFileTree(dir, object : SimpleFileVisitor<Path>() {
```

Comment:
:x: `SimpleFileVisitor<Path>` has different behaviour than the original implementation for the `visitFileFailed` and `postVisitDirectory` cases. In the original Groovy implementation the exceptions were being ignored. `SimpleFileVisitor<Path>` will throw.

Reply:
Good catch! Fixed